### PR TITLE
layered_repo - atomic last/prev record_lsn

### PR DESCRIPTION
**Pull request by patins**
Make a new type that stores both Lsns. Use an RwLock for thread safety.

First pass at #446

### CODE REVIEW

File: pageserver/src/layered_repository.rs [hunk 1]
```diff
@@ -1139,8 +1155,11 @@ impl LayeredTimeline {
     fn checkpoint_internal(&self, force: bool) -> Result<()> {
         // FIXME: these should be fetched atomically.
```
Comment (hlinnaka) [inline pageserver/src/layered_repository.rs:1156]: FIXME comment can be removed now.

Hmm, we probably should be reading last/prev_record_lsn first, and last_valid_lsn only after that. In general, `self.last_valid_lsn` can be ahead of `self.last_record_lsn`, but not the other way round. I'm not sure if anything bad would happen if they are the other way round in these local variables, but it feels wrong.

File: pageserver/src/layered_repository.rs [hunk 2]
```diff
@@ -829,12 +849,7 @@ impl Timeline for LayeredTimeline {
...
-        // Can't move backwards.
-        let old = self.last_record_lsn.fetch_max(lsn);
-        assert!(old <= lsn);
-
-        // Use old value of last_record_lsn as prev_record_lsn
-        self.prev_record_lsn.fetch_max(old);
+        self.record_lsn.write().unwrap().advance(lsn);
 
         // Also advance last_valid_lsn
         let old = self.last_valid_lsn.advance(lsn);
```
Comment (patins) [inline pageserver/src/layered_repository.rs:855]: To satisfy the invariant `last_valid_lsn >= last_record_lsn`, I think we need the following orderings:

Loads (as you suggested):
1. `record_lsn`
2. `valid_lsn`

Advances:
1. `valid_lsn`
2. `record_lsn`

We should re-order the advances here (in `advance_last_record_lsn`) to match `init_valid_lsn`

File: pageserver/src/layered_repository.rs [hunk 3]
```diff
@@ -475,14 +490,16 @@ pub struct LayeredTimeline {
...
     //
     // We also remember the starting point of the previous record in
-    // 'prev_record_lsn'. It's used to set the xl_prev pointer of the
+    // 'record_lsn.prev'. It's used to set the xl_prev pointer of the
     // first WAL record when the node is started up. But here, we just
-    // keep track of it. FIXME: last_record_lsn and prev_record_lsn
-    // should be updated atomically together.
+    // keep track of it.
     //
+    // When advancing last_valid_lsn and record_lsn simaltaneously, we MUST
```
Comment (hlinnaka) [inline pageserver/src/layered_repository.rs:497]: typo: simultaneously

**Review by hlinnaka — APPROVED** [review]

File: pageserver/src/layered_repository.rs (hunk 3 above)
Comment (patins) [reply pageserver/src/layered_repository.rs]: Whoops

**Pull request merged**